copy/copy.go

```diff
@@ -410,7 +414,34 @@ func (c *copier) copyMultipleImages(ctx context.Context, policyContext *signatur
 			return nil, errors.Wrapf(err, "Can not copy signatures to %s", transports.ImageName(c.dest.Reference()))
 		}
 	}
-	canModifyManifestList := (len(sigs) == 0)
+
+	// If the destination is a digested reference, make a note of that, determine what digest value we're
+	// expecting, and check that the source manifest matches it.
+	destIsDigestedReference := false
+	if named := c.dest.Reference().DockerReference(); named != nil {
+		if digested, ok := named.(reference.Digested); ok {
+			destIsDigestedReference = true
+			matches, err := manifest.MatchesDigest(manifestList, digested.Digest())
+			if err != nil {
+				return nil, errors.Wrapf(err, "computing digest of source image's manifest")
+			}
+			if !matches {
+				return nil, errors.New("Digest of source image's manifest would not match destination reference")
+			}
+		}
+	}
+
+	// Determine if we're allowed to modify the manifest and why not
+	canModifyManifestList := ""
+	if len(sigs) > 0 {
+		canModifyManifestList = "Would invalidate signatures"
+	}
+	if destIsDigestedReference {
+		canModifyManifestList = "Destination specifies a digest"
+	}
+	if options.PreserveDigests {
+		canModifyManifestList = "Digests are being preserved"
```

Comment:
Maybe something like “Instructed to preserve digests”, to point at some caller?

I’m not at all sure this is much better, we still can’t tell the user what this means exactly (e.g. what CLI option that was).

(This should be the same in both cases.)